torch/csrc/jit/codegen/cuda/runtime/memory.cu

```diff
@@ -0,0 +1,107 @@
+// Utility macro for this file
+#define DEVICE_INLINE __device__ inline
+
+#if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 750))
+
+namespace Turing {
+
+namespace util {
+
+// Special utility for ldmatrix and cp_async
+DEVICE_INLINE unsigned toSmem(const void* ptr) {
+  unsigned smem_ptr;
+
+  asm("{ .reg .u64 smem_ptr; cvta.to.shared.u64 smem_ptr, %1; cvt.u32.u64 %0, smem_ptr; }\n"
+      : "=r"(smem_ptr)
+      : "l"(ptr));
+
+  return smem_ptr;
+}
+
+} // namespace util
+
+DEVICE_INLINE void ldMatrix(Array<__half, 4, 4>& out, void const* ptr) {
```

Comment:
Load Matrix (per warp instruction) is to take data from SMEM to Local Memory. Automatically handles vectorized loads/stores in the MMA operation. Loads 8x8 matrix into a warp. Thread 0-7 provide the ptr that is the start of each row. All other threads can simply point to something valid (including 0).

The x2 modifier on the instruction will actually load 2x8 rows to make a 16x8, then thread 0-15 will specify the start of each row.

Finally is an x4 modifier producing a 32x8 using addrs from 0-31 in each warp.

Reply:
Added to comment. Thanks.